README.md
# torrent-stream-vlc
<i>Insert magnet link or torrent file for streaming directly to VLC (VideoLAN Client) on Mac OS X</i>

Built using <a href="https://www.npmjs.com/package/torrent-stream">torrent-stream</a> and inspired by <a href="https://www.npmjs.com/package/peerflix">peerflix</a>. Thanks to <a href="https://github.com/mafintosh">Mathias  Buus.</a>

## Prerequisites
- node + npm 
- VLC for Max OS X (https://www.videolan.org/vlc/download-macosx.html)

## Limitations
Due to time and testing constraints, this package was developed only for systems running OS X with VLC player installed. Multiplatform support as well as streaming targets other than VLC might be added later on (If you want to do this yourself, have a closer look at the peerflix package mentioned above).

## Usage

### Initialise

    // import the module
    var TorrentStreamVLC = require('torrent-stream-vlc');

    // create an instance
    var torrentStreamVLC = new TorrentStreamVLC.default();

### List files included in torrent and start streaming

    // call getFileList with torrent (can be the actual torrent file or a magnet link) to output a list of choices (array) including e.g. file name and index
    torrentStreamVLC.getFileList(torrent).then(function(choices) {
      // here you need to select the file to target (video file)
      var fileIndex = 1 // the index of the targeted file

      // start the stream with target index
      torrentStreamVLC.startStream(fileIndex)
    })

### Stop and delete the currently running stream

    // removes and destroys torrent-stream engine, emits 'stream-aborted' event
    torrentStreamVLC.destroyTorrent()

### Listen to events

    // receive notification when stream is ready
    torrentStreamVLC.on('stream-ready', function(info) {
      /* info contains: 
        href (address of the stream server)
        storagePath (path to where stream data is temporarily saved)
      */
    })

    // get stream status updates every 500 ms
    torrentStreamVLC.on('stream-status', function(status) {
      /* status contains: 
        downloadSpeed (current download speed)
        uploadSpeed (current upload speed)
        uploadedSize (current upload speed)
        downloadedSize (size of already downloaded pieces)
        downloadedPercentage (percent of file downloaded)
        peersActive (connected peers)
        peersAvailable (available peers)
        streamRuntime (time the stream is already running in seconds)
        hotswaps (hotswap occurences)
        verified (verified pieces)
        invalid (invalid pieces)
        queued (queued peers)
      */
    })

    // stream-abort happens when vlc was exited
    torrentStreamVLC.on('stream-aborted', function() {
      // e.g. exit the process
      process.exit(0)
    })

## Contribution & Modification
Contribute to this work by improving it or modify the package for your own needs. I recommend working with the included src file and babel:

    // install babel-cli and babel-preset-es2015
    npm run install-dev

    // run this to watch and compile new dist file
    npm run dev

CHANGELOG.md
## <sub>v0.1.4</sub>
##### Jan. 23, 2017
* Improvements and bug fixes

## <sub>v0.1.0</sub>
##### Jan. 20, 2017
* Initial release

torrent-stream-vlc.src.js
import torrentStream from 'torrent-stream'
import http from 'http'
import fs from 'fs'
import rangeParser from 'range-parser'
import url from 'url'
import mime from 'mime'
import pump from 'pump'
import numeral from 'numeral'
import { exec } from 'child_process'
import EventEmitter from 'events'

class TorrentStreamVLC extends EventEmitter {
  constructor(props) {
    super()
  }

  // start stream with torrent magnet link
  startStream(index) {
    this.engine = torrentStream(this.torrent)
    this.engine.on('ready', () => {
      this.engine.server = this.createStreamServer(index)
      this.engine.server.listen(0, 'localhost')
      this.engine.listen()

      let hotswaps = 0
      let verified = 0
      let invalid = 0
      let downloadedPercentage = 0
      let streamStartTime = Date.now()

      this.engine.on('verify', () => {
        verified++

        if (this.engine.torrent) {
          let streamLength = 0
          let torrent = this.engine.torrent
          if (this.engine.torrent.files.length > 0) {
            let difference = torrent.length - torrent.files[index].length
            streamLength = torrent.length - difference
          } else {
            streamLength = torrent.length
          }
          downloadedPercentage = Math.floor(this.engine.swarm.downloaded * 100 / streamLength)
        }
      })
      this.engine.on('invalid-piece', () => {
        invalid++
      })
      this.engine.on('hotswap', () => {
        hotswaps++
      })
      this.engine.server.on('listening', () => {
        // this.engine.files[index].select()        
        this.startVLC()
        const logStatus = () => {
          let unchoked = this.engine.swarm.wires.filter(wire => {
            return !wire.peerChoking
          })
          let streamRuntime = Math.floor((Date.now() - streamStartTime) / 1000)

          let status = {
            downloadSpeed: this.getFormattedByteNumber(this.engine.swarm.downloadSpeed()),
            uploadSpeed: this.getFormattedByteNumber(this.engine.swarm.uploadSpeed()),
            downloadedSize: this.getFormattedByteNumber(this.engine.swarm.downloaded),
            uploadedSize: this.getFormattedByteNumber(this.engine.swarm.uploaded),
            downloadedPercentage,
            peersActive: unchoked.length,
            peersAvailable: this.engine.swarm.wires.length,
            streamRuntime,
            hotswaps,
            verified,
            invalid,
            queued: this.engine.swarm.queued
          }
          this.emit('stream-status', status)
        }

        this.interval = setInterval(logStatus, 500)
        logStatus()
      })

      this.engine.server.once('error', () => {
        this.engine.server.listen(0, 'localhost')
      })
    })
  }

  // format byte number for readable log output
  getFormattedByteNumber(number) {
    return numeral(number).format('0.0b')
  }

  // get a list of possible stream targets
  getFileList(torrent) {
    return new Promise((resolve, reject) => {
      this.torrent = torrent
      let isTorrent = false
      if (this.torrent.indexOf('magnet:?') > -1) {
        isTorrent = true
      } else if (/(?:\.([^.]+))?$/.exec(this.torrent)[1] == 'torrent') {
        isTorrent = true
        this.torrent = fs.readFileSync(torrent)
      }

      if (isTorrent) {
        this.engine = torrentStream(this.torrent);
        this.engine.on('ready', () => {
          let choices = this.engine.files.map((file, index) => {
            return {
              name: file.name,
              size: this.getFormattedByteNumber(file.length),
              value: index
            }
          })
          this.engine.destroy(() => {
            resolve(choices)
          })
        })
      } else {
        reject('Invalid input: no magnet link or torrent was inserted')
      }
    });
  }

  // destroys engine, removes files, exits stream process
  destroyTorrent() {
    if (this.interval) clearInterval(this.interval)
    if (this.engine.server) this.engine.server.close()
    this.engine.remove(() => {
      this.engine.destroy(() => {
        this.emit('stream-aborted')
      })
    })
  }

  // start VLC process pointing to stream server
  startVLC() {
    const href = 'http://localhost:' + this.engine.server.address().port + '/.m3u'
    const root = '/Applications/VLC.app/Contents/MacOS/VLC'
    const home = (process.env.HOME || '') + root
    const vlc = exec('vlc --video-on-top --play-and-exit ' + href + ' || ' + root + ' ' + href, (error, stdout, stderror) => {
      if (error) process.exit(0)
    })

    vlc.on('exit', () => {
      this.destroyTorrent()
    })

    this.emit('stream-ready', {
      href, 
      storagePath: this.engine.path
    })
  }

  // creates a stream server for vlc requests
  createStreamServer(index) {
    let server = http.createServer()
    let getType = mime.lookup.bind(mime)
    server.on('request', (request, response) => {
      let u = url.parse(request.url)
      let host = 'localhost'
      let file = this.engine.files[index]
      let range = request.headers.range

      let toJSON = () => {
        let totalPeers = this.engine.swarm.wires
        let activePeers = totalPeers.filter(wire => {
          return !wire.peerChoking
        })

        let swarmStats = {
          downloaded: this.engine.swarm.downloaded,
          uploaded: this.engine.swarm.uploaded,
          downloadSpeed: parseInt(this.engine.swarm.downloadSpeed(), 10),
          uploadSpeed: parseInt(this.engine.swarm.uploadSpeed(), 10),
          totalPeers: totalPeers.length,
          activePeers: activePeers.length
        }

        return JSON.stringify(swarmStats, null, '  ')
      }

      if (request.method === 'OPTIONS' && request.headers['access-control-request-headers']) {
        response.setHeader('Access-Control-Allow-Origin', request.headers.origin)
        response.setHeader('Access-Control-Allow-Methods', 'POST, GET, OPTIONS')
        response.setHeader(
            'Access-Control-Allow-Headers',
            request.headers['access-control-request-headers'])
        response.setHeader('Access-Control-Max-Age', '1728000')

        response.end()
        return
      }

      if (request.headers.origin) response.setHeader('Access-Control-Allow-Origin', request.headers.origin)

      if (u.pathname === '/favicon.ico') {
        response.statusCode = 404
        response.end()
        return
      }

      if (u.pathname === '/.json') {
        let json = toJSON()
        response.setHeader('Content-Type', 'application/json; charset=utf-8')
        response.setHeader('Content-Length', Buffer.byteLength(json))
        response.end(json)
        return
      }

      if (u.pathname === '/.m3u') {
        let playlist = '#EXTM3U\n#EXTINF:-1,' + file.path + '\n' + 'http://localhost:' + this.engine.server.address().port
        response.setHeader('Content-Type', 'application/x-mpegurl; charset=utf-8')
        response.setHeader('Content-Length', Buffer.byteLength(playlist))
        response.end(playlist)
        return
      }

      range = range && rangeParser(file.length, range)[0]
      response.setHeader('Accept-Ranges', 'bytes')
      response.setHeader('Content-Type', getType(file.name))
      response.setHeader('transferMode.dlna.org', 'Streaming')
      response.setHeader('contentFeatures.dlna.org', 'DLNA.ORG_OP=01;DLNA.ORG_CI=0;DLNA.ORG_FLAGS=017000 00000000000000000000000000')
      if (!range) {
        response.setHeader('Content-Length', file.length)
        if (request.method === 'HEAD') return response.end()
        pump(file.createReadStream(), response)
        return
      }

      response.statusCode = 206
      response.setHeader('Content-Length', range.end - range.start + 1)
      response.setHeader('Content-Range', 'bytes ' + range.start + '-' + range.end + '/' + file.length)
      if (request.method === 'HEAD') return response.end()
      pump(file.createReadStream(range), response)
    })

    server.on('connection', socket => {
      socket.setTimeout(36000000)
    })

    return server
  }
}

export default TorrentStreamVLC

vlc-torrent-stream.src.js
import torrentStream from 'torrent-stream'
import http from 'http'
import fs from 'fs'
import rangeParser from 'range-parser'
import url from 'url'
import mime from 'mime'
import pump from 'pump'
import numeral from 'numeral'
import util from 'util'
import { exec } from 'child_process'
import EventEmitter from 'events'

class VlcTorrentStream extends EventEmitter {
  constructor(props) {
    super()
    this.torrent = props
  }

  // start stream with torrent magnet link
  startStream(index) {
    this.engine = torrentStream(this.torrent)
    this.engine.on('ready', () => {
      this.engine.server = this.createStreamServer(index)
      this.engine.server.listen(0, 'localhost')
      this.engine.listen()

      let hotswaps = 0
      let verified = 0
      let invalid = 0
      let downloadedPercentage = 0
      let streamStartTime = Date.now()

      this.engine.on('verify', () => {
        verified++
        downloadedPercentage = Math.floor(verified / this.engine.torrent.pieces.length * 100)
      })
      this.engine.on('invalid-piece', () => {
        invalid++
      })
      this.engine.on('hotswap', () => {
        hotswaps++
      })
      this.engine.server.on('listening', () => {
        this.engine.files[index].select()        
        this.startVLC()
        const logStatus = () => {
          let unchoked = this.engine.swarm.wires.filter(wire => {
            return !wire.peerChoking
          })
          let streamRuntime = Math.floor((Date.now() - streamStartTime) / 1000)

          let status = {
            downloadSpeed: this.getFormattedByteNumber(this.engine.swarm.downloadSpeed()),
            downloadedSize: this.getFormattedByteNumber(this.engine.swarm.downloaded),
            uploadedSize: this.getFormattedByteNumber(this.engine.swarm.uploaded),
            downloadedPercentage,
            peersActive: unchoked.length,
            peersAvailable: this.engine.swarm.wires.length,
            streamRuntime,
            hotswaps,
            verified,
            invalid,
            queued: this.engine.swarm.queued
          }
          this.emit('stream-status', status)
        }

        setInterval(logStatus, 500)
        logStatus()
      })

      this.engine.server.once('error', () => {
        this.engine.server.listen(0, 'localhost')
      })
    })
  }

  // format byte number for readable log output
  getFormattedByteNumber(number) {
    return numeral(number).format('0.0b')
  }

  // get a list of possible stream targets
  getFileList() {
    return new Promise((resolve) => {
      this.engine = torrentStream(this.torrent)
      this.engine.on('ready', () => {
        let choices = this.engine.files.map((file, index) => {
          return {
            name: file.name,
            size: this.getFormattedByteNumber(file.length),
            value: index
          }
        })
        resolve(choices)
      })
    });
  }

  // destroys engine, removes files, exits stream process
  exitAndRemove() {
    this.engine.remove(() => {
      this.emit('stream-abort')
    })
  }

  // start VLC process pointing to stream server
  startVLC() {
    const href = 'http://localhost:' + this.engine.server.address().port + '/.m3u'
    const root = '/Applications/VLC.app/Contents/MacOS/VLC'
    const home = (process.env.HOME || '') + root
    const vlc = exec('vlc --video-on-top --play-and-exit ' + href + ' || ' + root + ' ' + href, (error, stdout, stderror) => {
      if (error) process.exit(0)
    })

    const _this = this
    vlc.on('exit', function () {
      _this.exitAndRemove()
    })

    this.emit('stream-ready', {
      href, 
      storagePath: this.engine.path
    })
  }

  // creates a stream server for vlc requests
  createStreamServer(index) {
    let server = http.createServer()
    let getType = mime.lookup.bind(mime)
    server.on('request', (request, response) => {
      let u = url.parse(request.url)
      let host = 'localhost'
      let file = this.engine.files[index]
      let range = request.headers.range

      let toJSON = () => {
        let totalPeers = this.engine.swarm.wires
        let activePeers = totalPeers.filter(function (wire) {
          return !wire.peerChoking
        })

        let swarmStats = {
          downloaded: this.engine.swarm.downloaded,
          uploaded: this.engine.swarm.uploaded,
          downloadSpeed: parseInt(this.engine.swarm.downloadSpeed(), 10),
          uploadSpeed: parseInt(this.engine.swarm.uploadSpeed(), 10),
          totalPeers: totalPeers.length,
          activePeers: activePeers.length
        }

        return JSON.stringify(swarmStats, null, '  ')
      }

      if (request.method === 'OPTIONS' && request.headers['access-control-request-headers']) {
        response.setHeader('Access-Control-Allow-Origin', request.headers.origin)
        response.setHeader('Access-Control-Allow-Methods', 'POST, GET, OPTIONS')
        response.setHeader(
            'Access-Control-Allow-Headers',
            request.headers['access-control-request-headers'])
        response.setHeader('Access-Control-Max-Age', '1728000')

        response.end()
        return
      }

      if (request.headers.origin) response.setHeader('Access-Control-Allow-Origin', request.headers.origin)

      if (u.pathname === '/favicon.ico') {
        response.statusCode = 404
        response.end()
        return
      }

      if (u.pathname === '/.json') {
        let json = toJSON()
        response.setHeader('Content-Type', 'application/json; charset=utf-8')
        response.setHeader('Content-Length', Buffer.byteLength(json))
        response.end(json)
        return
      }

      if (u.pathname === '/.m3u') {
        let playlist = '#EXTM3U\n#EXTINF:-1,' + file.path + '\n' + 'http://localhost:' + this.engine.server.address().port
        response.setHeader('Content-Type', 'application/x-mpegurl; charset=utf-8')
        response.setHeader('Content-Length', Buffer.byteLength(playlist))
        response.end(playlist)
        return
      }

      range = range && rangeParser(file.length, range)[0]
      response.setHeader('Accept-Ranges', 'bytes')
      response.setHeader('Content-Type', getType(file.name))
      response.setHeader('transferMode.dlna.org', 'Streaming')
      response.setHeader('contentFeatures.dlna.org', 'DLNA.ORG_OP=01;DLNA.ORG_CI=0;DLNA.ORG_FLAGS=017000 00000000000000000000000000')
      if (!range) {
        response.setHeader('Content-Length', file.length)
        if (request.method === 'HEAD') return response.end()
        pump(file.createReadStream(), response)
        return
      }

      response.statusCode = 206
      response.setHeader('Content-Length', range.end - range.start + 1)
      response.setHeader('Content-Range', 'bytes ' + range.start + '-' + range.end + '/' + file.length)
      if (request.method === 'HEAD') return response.end()
      pump(file.createReadStream(range), response)
    })

    server.on('connection', socket => {
      socket.setTimeout(36000000)
    })

    return server
  }
}

export default VlcTorrentStream

torrent-stream-vlc.dist.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _torrentStream = require('torrent-stream');

var _torrentStream2 = _interopRequireDefault(_torrentStream);

var _http = require('http');

var _http2 = _interopRequireDefault(_http);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _rangeParser = require('range-parser');

var _rangeParser2 = _interopRequireDefault(_rangeParser);

var _url = require('url');

var _url2 = _interopRequireDefault(_url);

var _mime = require('mime');

var _mime2 = _interopRequireDefault(_mime);

var _pump = require('pump');

var _pump2 = _interopRequireDefault(_pump);

var _numeral = require('numeral');

var _numeral2 = _interopRequireDefault(_numeral);

var _child_process = require('child_process');

var _events = require('events');

var _events2 = _interopRequireDefault(_events);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var TorrentStreamVLC = function (_EventEmitter) {
  _inherits(TorrentStreamVLC, _EventEmitter);

  function TorrentStreamVLC(props) {
    _classCallCheck(this, TorrentStreamVLC);

    return _possibleConstructorReturn(this, (TorrentStreamVLC.__proto__ || Object.getPrototypeOf(TorrentStreamVLC)).call(this));
  }

  // start stream with torrent magnet link


  _createClass(TorrentStreamVLC, [{
    key: 'startStream',
    value: function startStream(index) {
      var _this2 = this;

      this.engine = (0, _torrentStream2.default)(this.torrent);
      this.engine.on('ready', function () {
        _this2.engine.server = _this2.createStreamServer(index);
        _this2.engine.server.listen(0, 'localhost');
        _this2.engine.listen();

        var hotswaps = 0;
        var verified = 0;
        var invalid = 0;
        var downloadedPercentage = 0;
        var streamStartTime = Date.now();

        _this2.engine.on('verify', function () {
          verified++;

          if (_this2.engine.torrent) {
            var streamLength = 0;
            var torrent = _this2.engine.torrent;
            if (_this2.engine.torrent.files.length > 0) {
              var difference = torrent.length - torrent.files[index].length;
              streamLength = torrent.length - difference;
            } else {
              streamLength = torrent.length;
            }
            downloadedPercentage = Math.floor(_this2.engine.swarm.downloaded * 100 / streamLength);
          }
        });
        _this2.engine.on('invalid-piece', function () {
          invalid++;
        });
        _this2.engine.on('hotswap', function () {
          hotswaps++;
        });
        _this2.engine.server.on('listening', function () {
          // this.engine.files[index].select()        
          _this2.startVLC();
          var logStatus = function logStatus() {
            var unchoked = _this2.engine.swarm.wires.filter(function (wire) {
              return !wire.peerChoking;
            });
            var streamRuntime = Math.floor((Date.now() - streamStartTime) / 1000);

            var status = {
              downloadSpeed: _this2.getFormattedByteNumber(_this2.engine.swarm.downloadSpeed()),
              uploadSpeed: _this2.getFormattedByteNumber(_this2.engine.swarm.uploadSpeed()),
              downloadedSize: _this2.getFormattedByteNumber(_this2.engine.swarm.downloaded),
              uploadedSize: _this2.getFormattedByteNumber(_this2.engine.swarm.uploaded),
              downloadedPercentage: downloadedPercentage,
              peersActive: unchoked.length,
              peersAvailable: _this2.engine.swarm.wires.length,
              streamRuntime: streamRuntime,
              hotswaps: hotswaps,
              verified: verified,
              invalid: invalid,
              queued: _this2.engine.swarm.queued
            };
            _this2.emit('stream-status', status);
          };

          _this2.interval = setInterval(logStatus, 500);
          logStatus();
        });

        _this2.engine.server.once('error', function () {
          _this2.engine.server.listen(0, 'localhost');
        });
      });
    }

    // format byte number for readable log output

  }, {
    key: 'getFormattedByteNumber',
    value: function getFormattedByteNumber(number) {
      return (0, _numeral2.default)(number).format('0.0b');
    }

    // get a list of possible stream targets

  }, {
    key: 'getFileList',
    value: function getFileList(torrent) {
      var _this3 = this;

      return new Promise(function (resolve, reject) {
        _this3.torrent = torrent;
        var isTorrent = false;
        if (_this3.torrent.indexOf('magnet:?') > -1) {
          isTorrent = true;
        } else if (/(?:\.([^.]+))?$/.exec(_this3.torrent)[1] == 'torrent') {
          isTorrent = true;
          _this3.torrent = _fs2.default.readFileSync(torrent);
        }

        if (isTorrent) {
          _this3.engine = (0, _torrentStream2.default)(_this3.torrent);
          _this3.engine.on('ready', function () {
            var choices = _this3.engine.files.map(function (file, index) {
              return {
                name: file.name,
                size: _this3.getFormattedByteNumber(file.length),
                value: index
              };
            });
            _this3.engine.destroy(function () {
              resolve(choices);
            });
          });
        } else {
          reject('Invalid input: no magnet link or torrent was inserted');
        }
      });
    }

    // destroys engine, removes files, exits stream process

  }, {
    key: 'destroyTorrent',
    value: function destroyTorrent() {
      var _this4 = this;

      if (this.interval) clearInterval(this.interval);
      if (this.engine.server) this.engine.server.close();
      this.engine.remove(function () {
        _this4.engine.destroy(function () {
          _this4.emit('stream-aborted');
        });
      });
    }

    // start VLC process pointing to stream server

  }, {
    key: 'startVLC',
    value: function startVLC() {
      var _this5 = this;

      var href = 'http://localhost:' + this.engine.server.address().port + '/.m3u';
      var root = '/Applications/VLC.app/Contents/MacOS/VLC';
      var home = (process.env.HOME || '') + root;
      var vlc = (0, _child_process.exec)('vlc --video-on-top --play-and-exit ' + href + ' || ' + root + ' ' + href, function (error, stdout, stderror) {
        if (error) process.exit(0);
      });

      vlc.on('exit', function () {
        _this5.destroyTorrent();
      });

      this.emit('stream-ready', {
        href: href,
        storagePath: this.engine.path
      });
    }

    // creates a stream server for vlc requests

  }, {
    key: 'createStreamServer',
    value: function createStreamServer(index) {
      var _this6 = this;

      var server = _http2.default.createServer();
      var getType = _mime2.default.lookup.bind(_mime2.default);
      server.on('request', function (request, response) {
        var u = _url2.default.parse(request.url);
        var host = 'localhost';
        var file = _this6.engine.files[index];
        var range = request.headers.range;

        var toJSON = function toJSON() {
          var totalPeers = _this6.engine.swarm.wires;
          var activePeers = totalPeers.filter(function (wire) {
            return !wire.peerChoking;
          });

          var swarmStats = {
            downloaded: _this6.engine.swarm.downloaded,
            uploaded: _this6.engine.swarm.uploaded,
            downloadSpeed: parseInt(_this6.engine.swarm.downloadSpeed(), 10),
            uploadSpeed: parseInt(_this6.engine.swarm.uploadSpeed(), 10),
            totalPeers: totalPeers.length,
            activePeers: activePeers.length
          };

          return JSON.stringify(swarmStats, null, '  ');
        };

        if (request.method === 'OPTIONS' && request.headers['access-control-request-headers']) {
          response.setHeader('Access-Control-Allow-Origin', request.headers.origin);
          response.setHeader('Access-Control-Allow-Methods', 'POST, GET, OPTIONS');
          response.setHeader('Access-Control-Allow-Headers', request.headers['access-control-request-headers']);
          response.setHeader('Access-Control-Max-Age', '1728000');

          response.end();
          return;
        }

        if (request.headers.origin) response.setHeader('Access-Control-Allow-Origin', request.headers.origin);

        if (u.pathname === '/favicon.ico') {
          response.statusCode = 404;
          response.end();
          return;
        }

        if (u.pathname === '/.json') {
          var json = toJSON();
          response.setHeader('Content-Type', 'application/json; charset=utf-8');
          response.setHeader('Content-Length', Buffer.byteLength(json));
          response.end(json);
          return;
        }

        if (u.pathname === '/.m3u') {
          var playlist = '#EXTM3U\n#EXTINF:-1,' + file.path + '\n' + 'http://localhost:' + _this6.engine.server.address().port;
          response.setHeader('Content-Type', 'application/x-mpegurl; charset=utf-8');
          response.setHeader('Content-Length', Buffer.byteLength(playlist));
          response.end(playlist);
          return;
        }

        range = range && (0, _rangeParser2.default)(file.length, range)[0];
        response.setHeader('Accept-Ranges', 'bytes');
        response.setHeader('Content-Type', getType(file.name));
        response.setHeader('transferMode.dlna.org', 'Streaming');
        response.setHeader('contentFeatures.dlna.org', 'DLNA.ORG_OP=01;DLNA.ORG_CI=0;DLNA.ORG_FLAGS=017000 00000000000000000000000000');
        if (!range) {
          response.setHeader('Content-Length', file.length);
          if (request.method === 'HEAD') return response.end();
          (0, _pump2.default)(file.createReadStream(), response);
          return;
        }

        response.statusCode = 206;
        response.setHeader('Content-Length', range.end - range.start + 1);
        response.setHeader('Content-Range', 'bytes ' + range.start + '-' + range.end + '/' + file.length);
        if (request.method === 'HEAD') return response.end();
        (0, _pump2.default)(file.createReadStream(range), response);
      });

      server.on('connection', function (socket) {
        socket.setTimeout(36000000);
      });

      return server;
    }
  }]);

  return TorrentStreamVLC;
}(_events2.default);

exports.default = TorrentStreamVLC;
